packages/graphql-mini-transforms/src/document.ts

```diff
@@ -28,17 +40,18 @@ export function cleanDocument(
     addTypename(definition);
   }
 
-  const normalizedSource = minifySource(print(document));
+  const documentSource = print(document);
+  const normalizedSource = minifySource(documentSource);
   const normalizedDocument = parse(normalizedSource);
 
   for (const definition of normalizedDocument.definitions) {
     stripLoc(definition);
   }
 
-  // This ID is a hash of the full file contents that are part of the document,
-  // including other documents that are injected in, but excluding any unused
-  // fragments. This is useful for things like persisted queries.
-  const id = createHash('sha256').update(normalizedSource).digest('hex');
+  const id =
+    generateId === undefined
```

Comment:
Instead of checking this here, can you add a default value when destructuring the argument?

Reply:
I'm not sure I follow this remark. What default value are you suggesting? The line you've commented on would still be required, so do you just want to explicitly destructure to `undefined` or `null`?

Reply:
I think what he proposed is something like : 
```ts
{removeUnused = true, generateId = defaultGenerateId}: CleanDocumentOptions = {},
...
const id = generateId(normalizedSource)
```

If we want to do something like this, both `defaultGenerateId` and `generateId` will have to receive the same source. Currently, `defaultGenerateId` receives the normalized source and the `generateId` receives the document source.

I don't remember why `generateId` needs the document source.

Reply:
Well, that's what the code originally looked like... the commits and comments leading to that change are still there for review. That's why I'm a bit confused.

While changing the code to prevent printing the source twice, I also changed it so that `minifySource` was only called once since the extra operations seemed to be a concern.

The general `generateId` needs the unminifed source (this is explained in a previous comment). The default implementation minifies it, so it was happening twice. That motivated that change.

I can change it back, but I'm going to address the other comments first and leave this for now, so I don't go in circles on this.

Reply:
Sorry if I just missed this, but I am a bit confused why `generateId` needs to take the "full" document source, not the minified version. I assumed everything should be operating on the minified source, since that includes all the relevant fragments and normalizes away differences in whitespace and commas.

Reply:
I think we need to take a decision on what we want here. Do we want to pass the minified version or the "full" document source to the `generateId` function? 

- Passing the minified version to the `generateId` would be the simplest solution but I think this was causing an issue for your use case @dsanders11 ?
- Passing the "full" document to both functions will result in minifying twice the document (not sure of the cost of this)
- Having the current implementation solves @dsanders11 limitation and ensures we only minify the source once.

The keypoint here is what is the limitation with the minified version @dsanders11 and does the cost of the modification is high.

Reply:
@alexandcote, I think you've accurately summarized where things stand.

> * Passing the minified version to the generateId would be the simplest solution but I think this was causing an issue for your use case @dsanders11 ?

In no way trying to be rude, but all of the previous discussion and explanation still exists in this PR, nothing has been deleted. If you don't remember something about certain choices I'd suggest looking back on the other comments here for clarification. [My comment from September](https://github.com/Shopify/graphql-tools-web/pull/99#discussion_r495517313) explains why the non-minified version is needed for `generateId`. If that explanation isn't clear, I can try to clarify any point of confusion. The TL;DR is simply that the string given to `generateId` should match the query sent to the server, and the query sent to the server is the non-minified version.

Reply:
Sorry, this is an unfortunate side effect of me/ us having let a PR review go for 4 months :p

I see where you are coming from on this, but I worry that the approach is very specific to one very special case where the printed, "normalized" document is identical to the original GraphQL source. This only happens when you aren't using fragments from external files, and you aren't letting the library automatically add `__typename` fields. IMO, this case is specific and uncommon enough (`__typename` additions are on by default), that we should push the pain of dealing with it to consumers' application code. An application can fairly easily run the same minification on the server source before computing the SHA (happy to export a utility for it if necessary), and that will work regardless of the fragment/ typename situation (and also works if you ever switch to our "simple" document format that doesn't need to be parsed and re-printed by the GraphQL client).

How do you feel about that?

Reply:
If we don't do that, I think we need to give the "generate me an ID" option three different sources that might be relevant for them: the minified source code, the original document source code (no fragments or typenames embedded), and the "normalized" document that includes all the fragments and typenames.